tests/src/Integration/Field/FieldTypeTest.php

```diff
@@ -0,0 +1,170 @@
+            'foreign keys' => [],
+        ], $storage->getSchema());
+    }
+
+    public function testProperties(): void
+    {
+        [$storage, ] = $this->createField('field_rgb', 'retrofit_field:field_example_rgb', 1);
+        $property = $storage->getPropertyDefinition('rgb');
+        self::assertNotNull($property);
+        self::assertEquals('string', $property->getDataType());
+    }
+
+    public function testIsEmpty(): void
+    {
+        $this->createField('field_rgb', 'retrofit_field:field_example_rgb', 1);
+        $entity = EntityTest::create();
+        self::assertTrue($entity->get('field_rgb')->isEmpty());
+        $entity->get('field_rgb')->setValue('#000000');
+        self::assertFalse($entity->get('field_rgb')->isEmpty());
+    }
+
+    public function testValidate(): void
+    {
+        $this->createField('field_rgb', 'retrofit_field:field_example_rgb', 1);
+        $entity = EntityTest::create();
+        $entity->get('field_rgb')->setValue('invalid');
+        $violations = $entity->get('field_rgb')->validate();
+        self::assertCount(1, $violations);
+        self::assertEquals(
+            'Color must be in the HTML format #abcdef.',
+            (string) $violations->get(0)->getMessage()
+        );
+        $entity->get('field_rgb')->setValue('#000000');
+        $violations = $entity->get('field_rgb')->validate();
+        self::assertCount(0, $violations);
+    }
+
+    public function testSavingEntity(): void
+    {
+        $this->installEntitySchema('entity_test');
+        $this->createField('field_rgb', 'retrofit_field:field_example_rgb', 1);
+        $entity = EntityTest::create();
+        $entity->get('field_rgb')->setValue('#000000');
+        $entity->save();
+
+        $database = $this->container->get('database');
+        self::assertInstanceOf(Connection::class, $database);
+        $schema = $database->schema();
+
+        self::assertTrue($schema->tableExists('entity_test__field_rgb'));
+        self::assertTrue($schema->fieldExists('entity_test__field_rgb', 'field_rgb_rgb'));
+
+        $values = $database->select('entity_test__field_rgb')
+            ->fields('entity_test__field_rgb', [])
+            ->execute()
+            ->fetchAll();
+        self::assertEquals('#000000', $values[0]->field_rgb_rgb);
+    }
+
+    /**
+     * @dataProvider fieldFormatterData
+     */
+    public function testFieldFormatter(string $plugin, string $output): void
+    {
+        $this->installEntitySchema('entity_test');
+        $this->createField('field_rgb', 'retrofit_field:field_example_rgb', 1);
+
+        $entity = EntityTest::create();
+        $entity->get('field_rgb')->setValue('#000000');
+        $entity->save();
+
+        $viewDisplay = EntityViewDisplay::collectRenderDisplay($entity, 'default');
+        $viewDisplay
+            ->setComponent('field_rgb', [
+                'type' => "retrofit_field_formatter:$plugin",
+            ])
+            ->save();
```

Comment:
This isn't saving the field formatter and always falls back to the default

Reply:
I'm not sure the `retrofit_field_formatter:` prefix is correct here. It may get added dynamically.